In [191]:
#import seaborn as sns
import csv
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.model_selection import StratifiedShuffleSplit,train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [2]:
def print_full(x):
    pd.set_option('display.max_rows', len(x))
    pd.set_option('display.max_columns', 500)
    pd.set_option('display.width', 2000)
    pd.set_option('display.float_format', '{:20,.2f}'.format)
    pd.set_option('display.max_colwidth', -1)
    print(x)
    pd.reset_option('display.max_rows')
    pd.reset_option('display.max_columns')
    pd.reset_option('display.width')
    pd.reset_option('display.float_format')
    pd.reset_option('display.max_colwidth')

In [3]:
data = pd.read_csv("weatherAUS.csv") 
data.head(10)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5  2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6  2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7  2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8  2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   
9  2008-12-10   Albury     13.1     30.1       1.4          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
0           W           44.0          W  ...        22.0       1007.7   
1         WNW           44.0        NNW  ...        25.0       1010.6   
2         WSW           46.0          W  ...        30.0       1007.6   
3          NE           24.0         SE  ...        16.0       1017.6   
4           W           41.0        ENE  ...        33.0       1010.8   
5         WNW           56.0          W  ...        23.0       1009.2   
6           W           50.0         SW  ...        19.0       1009.6   
7           W           35.0        SSE  ...        19.0       1013.4   
8         NNW           80.0         SE  ...         9.0       1008.9   
9           W           28.0          S  ...        27.0       1007.0   

   Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
0       1007.1       8.0       NaN     16.9     21.8         No      0.0   
1       1007.8       NaN       NaN     17.2     24.3         No      0.0   
2       1008.7       NaN       2.0     21.0     23.2         No      0.0   
3       1012.8       NaN       NaN     18.1     26.5         No      1.0   
4       1006.0       7.0       8.0     17.8     29.7         No      0.2   
5       1005.4       NaN       NaN     20.6     28.9         No      0.0   
6       1008.2       1.0       NaN     18.1     24.6         No      0.0   
7       1010.1       NaN       NaN     16.3     25.5         No      0.0   
8       1003.6       NaN       NaN     18.3     30.2         No      1.4   
9       1005.7       NaN       NaN     20.1     28.2        Yes      0.0   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  
5            No  
6            No  
7            No  
8           Yes  
9            No  

[10 rows x 24 columns]

In [4]:
data

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
0       2008-12-01   Albury     13.4     22.9       0.6          NaN   
1       2008-12-02   Albury      7.4     25.1       0.0          NaN   
2       2008-12-03   Albury     12.9     25.7       0.0          NaN   
3       2008-12-04   Albury      9.2     28.0       0.0          NaN   
4       2008-12-05   Albury     17.5     32.3       1.0          NaN   
...            ...      ...      ...      ...       ...          ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0          NaN   
142189  2017-06-21    Uluru      2.8     23.4       0.0          NaN   
142190  2017-06-22    Uluru      3.6     25.3       0.0          NaN   
142191  2017-06-23    Uluru      5.4     26.9       0.0          NaN   
142192  2017-06-24    Uluru      7.8     27.0       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
0            NaN           W           44.0          W  ...        22.0   
1            NaN         WNW           44.0        NNW  ...        25.0   
2            NaN         WSW           46.0          W  ...        30.0   
3            NaN          NE           24.0         SE  ...        16.0   
4            NaN           W           41.0        ENE  ...        33.0   
...          ...         ...            ...        ...  ...         ...   
142188       NaN           E           31.0        ESE  ...        27.0   
142189       NaN           E           31.0         SE  ...        24.0   
142190       NaN         NNW           22.0         SE  ...        21.0   
142191       NaN           N           37.0         SE  ...        24.0   
142192       NaN          SE           28.0        SSE  ...        24.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
142188       1024.7       1021.2       NaN       NaN      9.4     20.9   
142189       1024.6       1020.3       NaN       NaN     10.1     22.4   
142190       1023.5       1019.1       NaN       NaN     10.9     24.5   
142191       1021.0       1016.8       NaN       NaN     12.5     26.1   
142192       1019.4       1016.5       3.0       2.0     15.1     26.0   

        RainToday  RISK_MM  RainTomorrow  
0              No      0.0            No  
1              No      0.0            No  
2              No      0.0            No  
3              No      1.0            No  
4              No      0.2            No  
...           ...      ...           ...  
142188         No      0.0            No  
142189         No      0.0            No  
142190         No      0.0            No  
142191         No      0.0            No  
142192         No      0.0            No  

[142193 rows x 24 columns]

In [5]:
Braki = data.isna().sum()  # suma brakujacych wartosci w kazdej z kolumn

In [6]:
Braki/142193 * 100   #procent  NaN

Date              0.000000
Location          0.000000
MinTemp           0.447983
MaxTemp           0.226453
Rainfall          0.988797
Evaporation      42.789026
Sunshine         47.692924
WindGustDir       6.561504
WindGustSpeed     6.519308
WindDir9am        7.041838
WindDir3pm        2.656952
WindSpeed9am      0.948007
WindSpeed3pm      1.849599
Humidity9am       1.247600
Humidity3pm       2.538803
Pressure9am       9.855619
Pressure3pm       9.832411
Cloud9am         37.735332
Cloud3pm         40.152469
Temp9am           0.635756
Temp3pm           1.917113
RainToday         0.988797
RISK_MM           0.000000
RainTomorrow      0.000000
dtype: float64

In [7]:
data = data.drop( labels=["Cloud3pm","Cloud9am","Sunshine","Evaporation","RISK_MM"], 
                     axis=1, inplace=False, errors='raise') #riskmm = rainfall, temu wyrzucamy

In [8]:
data #usuniecie kolumn

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0       2008-12-01   Albury     13.4     22.9       0.6           W   
1       2008-12-02   Albury      7.4     25.1       0.0         WNW   
2       2008-12-03   Albury     12.9     25.7       0.0         WSW   
3       2008-12-04   Albury      9.2     28.0       0.0          NE   
4       2008-12-05   Albury     17.5     32.3       1.0           W   
...            ...      ...      ...      ...       ...         ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0           E   
142189  2017-06-21    Uluru      2.8     23.4       0.0           E   
142190  2017-06-22    Uluru      3.6     25.3       0.0         NNW   
142191  2017-06-23    Uluru      5.4     26.9       0.0           N   
142192  2017-06-24    Uluru      7.8     27.0       0.0          SE   

        WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0                44.0          W        WNW          20.0          24.0   
1                44.0        NNW        WSW           4.0          22.0   
2                46.0          W        WSW          19.0          26.0   
3                24.0         SE          E          11.0           9.0   
4                41.0        ENE         NW           7.0          20.0   
...               ...        ...        ...           ...           ...   
142188           31.0        ESE          E          15.0          13.0   
142189           31.0         SE        ENE          13.0          11.0   
142190           22.0         SE          N          13.0           9.0   
142191           37.0         SE        WNW           9.0           9.0   
142192           28.0        SSE          N          13.0           7.0   

        Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
0              71.0         22.0       1007.7       1007.1     16.9     21.8   
1              44.0         25.0       1010.6       1007.8     17.2     24.3   
2              38.0         30.0       1007.6       1008.7     21.0     23.2   
3              45.0         16.0       1017.6       1012.8     18.1     26.5   
4              82.0         33.0       1010.8       1006.0     17.8     29.7   
...             ...          ...          ...          ...      ...      ...   
142188         59.0         27.0       1024.7       1021.2      9.4     20.9   
142189         51.0         24.0       1024.6       1020.3     10.1     22.4   
142190         56.0         21.0       1023.5       1019.1     10.9     24.5   
142191         53.0         24.0       1021.0       1016.8     12.5     26.1   
142192         51.0         24.0       1019.4       1016.5     15.1     26.0   

       RainToday RainTomorrow  
0             No           No  
1             No           No  
2             No           No  
3             No           No  
4             No           No  
...          ...          ...  
142188        No           No  
142189        No           No  
142190        No           No  
142191        No           No  
142192        No           No  

[142193 rows x 19 columns]

In [9]:
data.head(50)

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0   2008-12-01   Albury     13.4     22.9       0.6           W   
1   2008-12-02   Albury      7.4     25.1       0.0         WNW   
2   2008-12-03   Albury     12.9     25.7       0.0         WSW   
3   2008-12-04   Albury      9.2     28.0       0.0          NE   
4   2008-12-05   Albury     17.5     32.3       1.0           W   
5   2008-12-06   Albury     14.6     29.7       0.2         WNW   
6   2008-12-07   Albury     14.3     25.0       0.0           W   
7   2008-12-08   Albury      7.7     26.7       0.0           W   
8   2008-12-09   Albury      9.7     31.9       0.0         NNW   
9   2008-12-10   Albury     13.1     30.1       1.4           W   
10  2008-12-11   Albury     13.4     30.4       0.0           N   
11  2008-12-12   Albury     15.9     21.7       2.2         NNE   
12  2008-12-13   Albury     15.9     18.6      15.6           W   
13  2008-12-14   Albury     12.6     21.0       3.6          SW   
14  2008-12-16   Albury      9.8     27.7       NaN         WNW   
15  2008-12-17   Albury     14.1     20.9       0.0         ENE   
16  2008-12-18   Albury     13.5     22.9      16.8           W   
17  2008-12-19   Albury     11.2     22.5      10.6         SSE   
18  2008-12-20   Albury      9.8     25.6       0.0         SSE   
19  2008-12-21   Albury     11.5     29.3       0.0           S   
20  2008-12-22   Albury     17.1     33.0       0.0          NE   
21  2008-12-23   Albury     20.5     31.8       0.0         WNW   
22  2008-12-24   Albury     15.3     30.9       0.0           N   
23  2008-12-25   Albury     12.6     32.4       0.0           W   
24  2008-12-26   Albury     16.2     33.9       0.0         WSW   
25  2008-12-27   Albury     16.9     33.0       0.0         WSW   
26  2008-12-28   Albury     20.1     32.7       0.0         WNW   
27  2008-12-29   Albury     19.7     27.2       0.0         WNW   
28  2008-12-30   Albury     12.5     24.2       1.2         WNW   
29  2008-12-31   Albury     12.0     24.4       0.8           W   
30  2009-01-01   Albury     11.3     26.5       0.0         WNW   
31  2009-01-02   Albury      9.6     23.9       0.0           W   
32  2009-01-03   Albury     10.5     28.8       0.0         SSE   
33  2009-01-04   Albury     12.3     34.6       0.0         WNW   
34  2009-01-05   Albury     12.9     35.8       0.0         WNW   
35  2009-01-06   Albury     13.7     37.9       0.0           W   
36  2009-01-07   Albury     16.1     38.9       0.0           W   
37  2009-01-08   Albury     14.0     28.3       0.0           W   
38  2009-01-09   Albury     12.5     28.4       0.0          NE   
39  2009-01-10   Albury     17.0     30.8       0.0          NE   
40  2009-01-11   Albury     16.9     32.0       0.0           S   
41  2009-01-12   Albury     17.3     34.7       0.0          SW   
42  2009-01-13   Albury     17.2     37.7       0.0         NNW   
43  2009-01-14   Albury     17.4     43.0       0.0          NW   
44  2009-01-15   Albury     19.8     32.7       0.0         WNW   
45  2009-01-16   Albury     14.9     26.7       0.0          SW   
46  2009-01-17   Albury     10.5     28.4       0.0          SE   
47  2009-01-18   Albury     11.3     32.2       0.0         WNW   
48  2009-01-19   Albury     13.9     36.6       0.0         WNW   
49  2009-01-20   Albury     18.6     39.9       0.0         NNW   

    WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0            44.0          W        WNW          20.0          24.0   
1            44.0        NNW        WSW           4.0          22.0   
2            46.0          W        WSW          19.0          26.0   
3            24.0         SE          E          11.0           9.0   
4            41.0        ENE         NW           7.0          20.0   
5            56.0          W          W          19.0          24.0   
6            50.0         SW          W          20.0          24.0   
7            35.0       

In [10]:
Braki = data.isna().sum()
Braki

Date                 0
Location             0
MinTemp            637
MaxTemp            322
Rainfall          1406
WindGustDir       9330
WindGustSpeed     9270
WindDir9am       10013
WindDir3pm        3778
WindSpeed9am      1348
WindSpeed3pm      2630
Humidity9am       1774
Humidity3pm       3610
Pressure9am      14014
Pressure3pm      13981
Temp9am            904
Temp3pm           2726
RainToday         1406
RainTomorrow         0
dtype: int64

# Imputancja

In [11]:
MinTemp_median         = data['MinTemp'].median()     #wyznaczenie miedian i dominant
MaxTemp_median         = data['MaxTemp'].median() 
Rainfall_median        = data['Rainfall'].median()
WindGustDir_dominant   = data['WindGustDir'].mode()
WindGustSpeed_median   = data['WindGustSpeed'].median()  
WindDir9am_dominant    = data['WindDir9am'].mode()   
WindDir3pm_dominant    = data['WindDir3pm'].mode()    
WindSpeed9am_median    = data['WindSpeed9am'].median() 
WindSpeed3pm_median    = data['WindSpeed3pm'].median() 
Humidity9am_median     = data['Humidity9am'].median()
Humidity3pm_median     = data['Humidity3pm'].median()
Pressure9am_median     = data['Pressure9am'].median()
Pressure3pm_median     = data['Pressure3pm'].median()
Temp9am_median         = data['Temp9am'].median()
Temp3pm_median         = data['Temp3pm'].median()
RainToday_dominant     =  data['RainToday'].mode() 

In [12]:
WindGustDir_dominant.to_numpy()[0]

'W'

In [13]:
WindGustSpeed_median 

39.0

In [14]:
data["MinTemp"].fillna(MinTemp_median, inplace = True) # uzupełnienie wartości
data['MaxTemp'].fillna(MaxTemp_median, inplace = True)
data['Rainfall'].fillna(Rainfall_median, inplace = True)
data['WindGustDir'].fillna(WindGustDir_dominant.to_numpy()[0], inplace = True)
data['WindGustSpeed'].fillna(WindGustSpeed_median, inplace = True)
data['WindDir9am'].fillna(WindDir9am_dominant.to_numpy()[0], inplace = True)
data['WindDir3pm'].fillna(WindDir3pm_dominant.to_numpy()[0], inplace = True)
data['WindSpeed9am'].fillna(WindSpeed9am_median, inplace = True)
data['WindSpeed3pm'].fillna(WindSpeed3pm_median, inplace = True)
data['Humidity9am'].fillna(Humidity9am_median, inplace = True)
data['Humidity3pm'].fillna(Humidity3pm_median, inplace = True)
data['Pressure9am'].fillna(Pressure9am_median, inplace = True)
data['Pressure3pm'].fillna(Pressure3pm_median, inplace = True)
data['Temp9am'].fillna(Temp9am_median, inplace = True)
data['Temp3pm'].fillna(Temp3pm_median, inplace = True)
data['RainToday'].fillna(RainToday_dominant.to_numpy()[0], inplace = True)

In [15]:
print_full(data['WindGustDir'][3460:3470])

3460    NNE
3461    E  
3462    WSW
3463    E  
3464    ENE
3465    W  
3466    ESE
3467    NNE
3468    NE 
3469    SSW
Name: WindGustDir, dtype: object


In [16]:
Braki = data.isna().sum()
Braki

Date             0
Location         0
MinTemp          0
MaxTemp          0
Rainfall         0
WindGustDir      0
WindGustSpeed    0
WindDir9am       0
WindDir3pm       0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Temp9am          0
Temp3pm          0
RainToday        0
RainTomorrow     0
dtype: int64

In [17]:
   def mod_outlier(df):                             #liczenie IQR
        df1 = df.copy()
        df = df._get_numeric_data()


        q1 = df.quantile(0.25)
        q3 = df.quantile(0.75)
        
        iqr = q3 - q1
        display(iqr)

        lower_bound = q1 -(1.5 * iqr) 
        upper_bound = q3 +(1.5 * iqr)


        for col in df.columns:
            for i in range(0,len(df[col])):
                if df[col][i] < lower_bound[col]:            
                    df[col][i] = 0.0000000000000000001

                if df[col][i] > upper_bound[col]:            
                    df[col][i] = 0.0000000000000000001    


        for col in df.columns:
            df1[col] = df[col]

        return(df1)

In [18]:
odcieta_data = mod_outlier(data)

MinTemp           9.2
MaxTemp          10.3
Rainfall          0.6
WindGustSpeed    15.0
WindSpeed9am     12.0
WindSpeed3pm     11.0
Humidity9am      26.0
Humidity3pm      28.0
Pressure9am       8.3
Pressure3pm       8.4
Temp9am           9.2
Temp3pm           9.6
dtype: float64

In [19]:
odcieta_data

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0       2008-12-01   Albury     13.4     22.9       0.6           W   
1       2008-12-02   Albury      7.4     25.1       0.0         WNW   
2       2008-12-03   Albury     12.9     25.7       0.0         WSW   
3       2008-12-04   Albury      9.2     28.0       0.0          NE   
4       2008-12-05   Albury     17.5     32.3       1.0           W   
...            ...      ...      ...      ...       ...         ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0           E   
142189  2017-06-21    Uluru      2.8     23.4       0.0           E   
142190  2017-06-22    Uluru      3.6     25.3       0.0         NNW   
142191  2017-06-23    Uluru      5.4     26.9       0.0           N   
142192  2017-06-24    Uluru      7.8     27.0       0.0          SE   

        WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0                44.0          W        WNW          20.0          24.0   
1                44.0        NNW        WSW           4.0          22.0   
2                46.0          W        WSW          19.0          26.0   
3                24.0         SE          E          11.0           9.0   
4                41.0        ENE         NW           7.0          20.0   
...               ...        ...        ...           ...           ...   
142188           31.0        ESE          E          15.0          13.0   
142189           31.0         SE        ENE          13.0          11.0   
142190           22.0         SE          N          13.0           9.0   
142191           37.0         SE        WNW           9.0           9.0   
142192           28.0        SSE          N          13.0           7.0   

        Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
0              71.0         22.0       1007.7       1007.1     16.9     21.8   
1              44.0         25.0       1010.6       1007.8     17.2     24.3   
2              38.0         30.0       1007.6       1008.7     21.0     23.2   
3              45.0         16.0       1017.6       1012.8     18.1     26.5   
4              82.0         33.0       1010.8       1006.0     17.8     29.7   
...             ...          ...          ...          ...      ...      ...   
142188         59.0         27.0       1024.7       1021.2      9.4     20.9   
142189         51.0         24.0       1024.6       1020.3     10.1     22.4   
142190         56.0         21.0       1023.5       1019.1     10.9     24.5   
142191         53.0         24.0       1021.0       1016.8     12.5     26.1   
142192         51.0         24.0       1019.4       1016.5     15.1     26.0   

       RainToday RainTomorrow  
0             No           No  
1             No           No  
2             No           No  
3             No           No  
4             No           No  
...          ...          ...  
142188        No           No  
142189        No           No  
142190        No           No  
142191        No           No  
142192        No           No  

[142193 rows x 19 columns]

In [20]:
odcieta_data= odcieta_data[odcieta_data.MinTemp != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.MaxTemp  != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Rainfall != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.WindGustSpeed != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.WindSpeed9am != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.WindSpeed3pm != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Humidity9am != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Humidity3pm  != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Pressure9am != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Pressure3pm != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Temp9am != 0.0000000000000000001 ]
odcieta_data = odcieta_data[odcieta_data.Temp3pm!= 0.0000000000000000001 ]

In [21]:
display(odcieta_data)  #df bez tych wystajacych nad quantile

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0       2008-12-01   Albury     13.4     22.9       0.6           W   
1       2008-12-02   Albury      7.4     25.1       0.0         WNW   
2       2008-12-03   Albury     12.9     25.7       0.0         WSW   
3       2008-12-04   Albury      9.2     28.0       0.0          NE   
4       2008-12-05   Albury     17.5     32.3       1.0           W   
...            ...      ...      ...      ...       ...         ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0           E   
142189  2017-06-21    Uluru      2.8     23.4       0.0           E   
142190  2017-06-22    Uluru      3.6     25.3       0.0         NNW   
142191  2017-06-23    Uluru      5.4     26.9       0.0           N   
142192  2017-06-24    Uluru      7.8     27.0       0.0          SE   

        WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0                44.0          W        WNW          20.0          24.0   
1                44.0        NNW        WSW           4.0          22.0   
2                46.0          W        WSW          19.0          26.0   
3                24.0         SE          E          11.0           9.0   
4                41.0        ENE         NW           7.0          20.0   
...               ...        ...        ...           ...           ...   
142188           31.0        ESE          E          15.0          13.0   
142189           31.0         SE        ENE          13.0          11.0   
142190           22.0         SE          N          13.0           9.0   
142191           37.0         SE        WNW           9.0           9.0   
142192           28.0        SSE          N          13.0           7.0   

        Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
0              71.0         22.0       1007.7       1007.1     16.9     21.8   
1              44.0         25.0       1010.6       1007.8     17.2     24.3   
2              38.0         30.0       1007.6       1008.7     21.0     23.2   
3              45.0         16.0       1017.6       1012.8     18.1     26.5   
4              82.0         33.0       1010.8       1006.0     17.8     29.7   
...             ...          ...          ...          ...      ...      ...   
142188         59.0         27.0       1024.7       1021.2      9.4     20.9   
142189         51.0         24.0       1024.6       1020.3     10.1     22.4   
142190         56.0         21.0       1023.5       1019.1     10.9     24.5   
142191         53.0         24.0       1021.0       1016.8     12.5     26.1   
142192         51.0         24.0       1019.4       1016.5     15.1     26.0   

       RainToday RainTomorrow  
0             No           No  
1             No           No  
2             No           No  
3             No           No  
4             No           No  
...          ...          ...  
142188        No           No  
142189        No           No  
142190        No           No  
142191        No           No  
142192        No           No  

[105507 rows x 19 columns]

In [22]:
kategorie = odcieta_data.loc[:,['Location',"WindGustDir",'WindDir9am','WindDir3pm','RainToday']]
kategorie

Location WindGustDir WindDir9am WindDir3pm RainToday
0        Albury           W          W        WNW        No
1        Albury         WNW        NNW        WSW        No
2        Albury         WSW          W        WSW        No
3        Albury          NE         SE          E        No
4        Albury           W        ENE         NW        No
...         ...         ...        ...        ...       ...
142188    Uluru           E        ESE          E        No
142189    Uluru           E         SE        ENE        No
142190    Uluru         NNW         SE          N        No
142191    Uluru           N         SE        WNW        No
142192    Uluru          SE        SSE          N        No

[105507 rows x 5 columns]

In [23]:
odcieta_data['Year']=[d.split('-')[0] for d in odcieta_data.Date]
odcieta_data['Month']=[d.split('-')[1] for d in odcieta_data.Date]
odcieta_data['Day']=[d.split('-')[2] for d in odcieta_data.Date]

In [24]:
odcieta_data = odcieta_data.drop( labels='Date', 
                     axis=1, inplace=False, errors='raise')
odcieta_data  # edycja daty rozdzielenie na year month day

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0        Albury     13.4     22.9       0.6           W           44.0   
1        Albury      7.4     25.1       0.0         WNW           44.0   
2        Albury     12.9     25.7       0.0         WSW           46.0   
3        Albury      9.2     28.0       0.0          NE           24.0   
4        Albury     17.5     32.3       1.0           W           41.0   
...         ...      ...      ...       ...         ...            ...   
142188    Uluru      3.5     21.8       0.0           E           31.0   
142189    Uluru      2.8     23.4       0.0           E           31.0   
142190    Uluru      3.6     25.3       0.0         NNW           22.0   
142191    Uluru      5.4     26.9       0.0           N           37.0   
142192    Uluru      7.8     27.0       0.0          SE           28.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  \
0               W        WNW          20.0          24.0  ...         22.0   
1             NNW        WSW           4.0          22.0  ...         25.0   
2               W        WSW          19.0          26.0  ...         30.0   
3              SE          E          11.0           9.0  ...         16.0   
4             ENE         NW           7.0          20.0  ...         33.0   
...           ...        ...           ...           ...  ...          ...   
142188        ESE          E          15.0          13.0  ...         27.0   
142189         SE        ENE          13.0          11.0  ...         24.0   
142190         SE          N          13.0           9.0  ...         21.0   
142191         SE        WNW           9.0           9.0  ...         24.0   
142192        SSE          N          13.0           7.0  ...         24.0   

        Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  \
0            1007.7       1007.1     16.9     21.8         No           No   
1            1010.6       1007.8     17.2     24.3         No           No   
2            1007.6       1008.7     21.0     23.2         No           No   
3            1017.6       1012.8     18.1     26.5         No           No   
4            1010.8       1006.0     17.8     29.7         No           No   
...             ...          ...      ...      ...        ...          ...   
142188       1024.7       1021.2      9.4     20.9         No           No   
142189       1024.6       1020.3     10.1     22.4         No           No   
142190       1023.5       1019.1     10.9     24.5         No           No   
142191       1021.0       1016.8     12.5     26.1         No           No   
142192       1019.4       1016.5     15.1     26.0         No           No   

        Year Month Day  
0       2008    12  01  
1       2008    12  02  
2       2008    12  03  
3       2008    12  04  
4       2008    12  05  
...      ...   ...  ..  
142188  2017    06  20  
142189  2017    06  21  
142190  2017    06  22  
142191  2017    06  23  
142192  2017    06  24  

[105507 rows x 21 columns]

In [25]:
odcieta_data

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0        Albury     13.4     22.9       0.6           W           44.0   
1        Albury      7.4     25.1       0.0         WNW           44.0   
2        Albury     12.9     25.7       0.0         WSW           46.0   
3        Albury      9.2     28.0       0.0          NE           24.0   
4        Albury     17.5     32.3       1.0           W           41.0   
...         ...      ...      ...       ...         ...            ...   
142188    Uluru      3.5     21.8       0.0           E           31.0   
142189    Uluru      2.8     23.4       0.0           E           31.0   
142190    Uluru      3.6     25.3       0.0         NNW           22.0   
142191    Uluru      5.4     26.9       0.0           N           37.0   
142192    Uluru      7.8     27.0       0.0          SE           28.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  \
0               W        WNW          20.0          24.0  ...         22.0   
1             NNW        WSW           4.0          22.0  ...         25.0   
2               W        WSW          19.0          26.0  ...         30.0   
3              SE          E          11.0           9.0  ...         16.0   
4             ENE         NW           7.0          20.0  ...         33.0   
...           ...        ...           ...           ...  ...          ...   
142188        ESE          E          15.0          13.0  ...         27.0   
142189         SE        ENE          13.0          11.0  ...         24.0   
142190         SE          N          13.0           9.0  ...         21.0   
142191         SE        WNW           9.0           9.0  ...         24.0   
142192        SSE          N          13.0           7.0  ...         24.0   

        Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  \
0            1007.7       1007.1     16.9     21.8         No           No   
1            1010.6       1007.8     17.2     24.3         No           No   
2            1007.6       1008.7     21.0     23.2         No           No   
3            1017.6       1012.8     18.1     26.5         No           No   
4            1010.8       1006.0     17.8     29.7         No           No   
...             ...          ...      ...      ...        ...          ...   
142188       1024.7       1021.2      9.4     20.9         No           No   
142189       1024.6       1020.3     10.1     22.4         No           No   
142190       1023.5       1019.1     10.9     24.5         No           No   
142191       1021.0       1016.8     12.5     26.1         No           No   
142192       1019.4       1016.5     15.1     26.0         No           No   

        Year Month Day  
0       2008    12  01  
1       2008    12  02  
2       2008    12  03  
3       2008    12  04  
4       2008    12  05  
...      ...   ...  ..  
142188  2017    06  20  
142189  2017    06  21  
142190  2017    06  22  
142191  2017    06  23  
142192  2017    06  24  

[105507 rows x 21 columns]

In [26]:
x = odcieta_data.loc[:,['MinTemp',"MaxTemp",'Rainfall','WindGustSpeed','WindSpeed9am',
                               'WindSpeed3pm','Humidity9am','Humidity3pm','Pressure9am',
                        'Pressure3pm','Temp9am','Temp3pm','Year','Month','Day']]

min_max_scaler = preprocessing.MinMaxScaler()
x_scaled = min_max_scaler.fit_transform(x)
normalized_data = pd.DataFrame(x_scaled, columns=x.columns)

In [27]:
normalized_data #znormalizowana data numeric

MinTemp   MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  \
0       0.538462  0.496333  0.400000       0.603448      0.540541   
1       0.373626  0.550122  0.000000       0.603448      0.108108   
2       0.524725  0.564792  0.000000       0.637931      0.513514   
3       0.423077  0.621027  0.000000       0.258621      0.297297   
4       0.651099  0.726161  0.666667       0.551724      0.189189   
...          ...       ...       ...            ...           ...   
105502  0.266484  0.469438  0.000000       0.379310      0.405405   
105503  0.247253  0.508557  0.000000       0.379310      0.351351   
105504  0.269231  0.555012  0.000000       0.224138      0.351351   
105505  0.318681  0.594132  0.000000       0.482759      0.243243   
105506  0.384615  0.596577  0.000000       0.327586      0.351351   

        WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  \
0           0.615385     0.646341     0.212121     0.199396     0.256716   
1           0.564103     0.317073     0.242424     0.287009     0.277612   
2           0.666667     0.243902     0.292929     0.196375     0.304478   
3           0.230769     0.329268     0.151515     0.498489     0.426866   
4           0.512821     0.780488     0.323232     0.293051     0.223881   
...              ...          ...          ...          ...          ...   
105502      0.333333     0.500000     0.262626     0.712991     0.677612   
105503      0.282051     0.402439     0.232323     0.709970     0.650746   
105504      0.230769     0.463415     0.202020     0.676737     0.614925   
105505      0.230769     0.426829     0.232323     0.601208     0.546269   
105506      0.179487     0.402439     0.232323     0.552870     0.537313   

         Temp9am   Temp3pm  Year     Month       Day  
0       0.501370  0.507812   0.1  1.000000  0.000000  
1       0.509589  0.572917   0.1  1.000000  0.033333  
2       0.613699  0.544271   0.1  1.000000  0.066667  
3       0.534247  0.630208   0.1  1.000000  0.100000  
4       0.526027  0.713542   0.1  1.000000  0.133333  
...          ...       ...   ...       ...       ...  
105502  0.295890  0.484375   1.0  0.454545  0.633333  
105503  0.315068  0.523437   1.0  0.454545  0.666667  
105504  0.336986  0.578125   1.0  0.454545  0.700000  
105505  0.380822  0.619792   1.0  0.454545  0.733333  
105506  0.452055  0.617187   1.0  0.454545  0.766667  

[105507 rows x 15 columns]

In [28]:
enc = preprocessing.OneHotEncoder()
enc.fit(kategorie)
onehotlabels = enc.transform(kategorie).toarray()
onehotlabels.shape

(105507, 99)

In [29]:
zakodowane= pd.DataFrame(onehotlabels)
zakodowane

0    1    2    3    4    5    6    7    8    9   ...   89   90   91  \
0       0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1       0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2       0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3       0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4       0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
105502  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
105503  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
105504  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
105505  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
105506  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

         92   93   94   95   96   97   98  
0       0.0  0.0  0.0  1.0  0.0  1.0  0.0  
1       0.0  0.0  0.0  0.0  1.0  1.0  0.0  
2       0.0  0.0  0.0  0.0  1.0  1.0  0.0  
3       0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4       0.0  0.0  0.0  0.0  0.0  1.0  0.0  
...     ...  ...  ...  ...  ...  ...  ...  
105502  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105503  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105504  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105505  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
105506  0.0  0.0  0.0  0.0  0.0  1.0  0.0  

[105507 rows x 99 columns]

In [30]:
abc = pd.DataFrame(odcieta_data.loc[:,'Location'].values, columns = ['Location'])
bca  = pd.DataFrame(odcieta_data.loc[:,'RainTomorrow'].values, columns = ['RainTomorrow'])
Big_boi = normalized_data.join(zakodowane)
Big_boi = abc.join(Big_boi)
Big_boi = bca.join(Big_boi)
Big_boi

RainTomorrow Location   MinTemp   MaxTemp  Rainfall  WindGustSpeed  \
0                No   Albury  0.538462  0.496333  0.400000       0.603448   
1                No   Albury  0.373626  0.550122  0.000000       0.603448   
2                No   Albury  0.524725  0.564792  0.000000       0.637931   
3                No   Albury  0.423077  0.621027  0.000000       0.258621   
4                No   Albury  0.651099  0.726161  0.666667       0.551724   
...             ...      ...       ...       ...       ...            ...   
105502           No    Uluru  0.266484  0.469438  0.000000       0.379310   
105503           No    Uluru  0.247253  0.508557  0.000000       0.379310   
105504           No    Uluru  0.269231  0.555012  0.000000       0.224138   
105505           No    Uluru  0.318681  0.594132  0.000000       0.482759   
105506           No    Uluru  0.384615  0.596577  0.000000       0.327586   

        WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  ...   89   90  \
0           0.540541      0.615385     0.646341     0.212121  ...  0.0  0.0   
1           0.108108      0.564103     0.317073     0.242424  ...  0.0  0.0   
2           0.513514      0.666667     0.243902     0.292929  ...  0.0  0.0   
3           0.297297      0.230769     0.329268     0.151515  ...  0.0  0.0   
4           0.189189      0.512821     0.780488     0.323232  ...  0.0  0.0   
...              ...           ...          ...          ...  ...  ...  ...   
105502      0.405405      0.333333     0.500000     0.262626  ...  0.0  0.0   
105503      0.351351      0.282051     0.402439     0.232323  ...  0.0  0.0   
105504      0.351351      0.230769     0.463415     0.202020  ...  0.0  0.0   
105505      0.243243      0.230769     0.426829     0.232323  ...  0.0  0.0   
105506      0.351351      0.179487     0.402439     0.232323  ...  0.0  0.0   

         91   92   93   94   95   96   97   98  
0       0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
1       0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  
2       0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  
3       0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4       0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
...     ...  ...  ...  ...  ...  ...  ...  ...  
105502  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105503  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105504  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105505  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
105506  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  

[105507 rows x 116 columns]

In [31]:
Big_boi = Big_boi.replace(list(set(Big_boi['RainTomorrow'])),\
                          [True, False]).astype({'RainTomorrow': 'bool'})
Big_boi["RainTomorrow"] = Big_boi["RainTomorrow"].astype(int)
Big_boi 

RainTomorrow Location   MinTemp   MaxTemp  Rainfall  WindGustSpeed  \
0                  1   Albury  0.538462  0.496333  0.400000       0.603448   
1                  1   Albury  0.373626  0.550122  0.000000       0.603448   
2                  1   Albury  0.524725  0.564792  0.000000       0.637931   
3                  1   Albury  0.423077  0.621027  0.000000       0.258621   
4                  1   Albury  0.651099  0.726161  0.666667       0.551724   
...              ...      ...       ...       ...       ...            ...   
105502             1    Uluru  0.266484  0.469438  0.000000       0.379310   
105503             1    Uluru  0.247253  0.508557  0.000000       0.379310   
105504             1    Uluru  0.269231  0.555012  0.000000       0.224138   
105505             1    Uluru  0.318681  0.594132  0.000000       0.482759   
105506             1    Uluru  0.384615  0.596577  0.000000       0.327586   

        WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  ...   89   90  \
0           0.540541      0.615385     0.646341     0.212121  ...  0.0  0.0   
1           0.108108      0.564103     0.317073     0.242424  ...  0.0  0.0   
2           0.513514      0.666667     0.243902     0.292929  ...  0.0  0.0   
3           0.297297      0.230769     0.329268     0.151515  ...  0.0  0.0   
4           0.189189      0.512821     0.780488     0.323232  ...  0.0  0.0   
...              ...           ...          ...          ...  ...  ...  ...   
105502      0.405405      0.333333     0.500000     0.262626  ...  0.0  0.0   
105503      0.351351      0.282051     0.402439     0.232323  ...  0.0  0.0   
105504      0.351351      0.230769     0.463415     0.202020  ...  0.0  0.0   
105505      0.243243      0.230769     0.426829     0.232323  ...  0.0  0.0   
105506      0.351351      0.179487     0.402439     0.232323  ...  0.0  0.0   

         91   92   93   94   95   96   97   98  
0       0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
1       0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  
2       0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  
3       0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4       0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
...     ...  ...  ...  ...  ...  ...  ...  ...  
105502  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105503  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105504  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
105505  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
105506  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  

[105507 rows x 116 columns]

In [32]:
def Regionowanko(df):
    df = df.copy()
    Regiony = np.unique(abc.values)
    lista_df = []
    for Region in Regiony:
        a = df.loc[df["Location"] == Region]
        lista_df.append(a)
    return lista_df

In [33]:
lista_region_df = Regionowanko(Big_boi)

In [83]:
len(lista_region_df)

49

In [34]:
lista_region_df[0]

RainTomorrow  Location   MinTemp   MaxTemp  Rainfall  WindGustSpeed  \
69267             1  Adelaide  0.519231  0.322738  0.533333       0.448276   
69268             1  Adelaide  0.340659  0.305623  0.000000       0.189655   
69269             1  Adelaide  0.315934  0.325183  0.000000       0.362069   
69270             0  Adelaide  0.480769  0.320293  0.000000       0.741379   
69271             0  Adelaide  0.428571  0.347188  0.000000       0.775862   
...             ...       ...       ...       ...       ...            ...   
71585             1  Adelaide  0.304945  0.403423  0.000000       0.172414   
71586             1  Adelaide  0.293956  0.405868  0.000000       0.137931   
71587             1  Adelaide  0.296703  0.413203  0.000000       0.224138   
71588             1  Adelaide  0.318681  0.388753  0.000000       0.172414   
71589             1  Adelaide  0.307692  0.359413  0.000000       0.137931   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  ...   89   90  \
69267      0.351351      0.384615     0.695122     0.515152  ...  0.0  0.0   
69268      0.054054      0.282051     0.768293     0.555556  ...  0.0  0.0   
69269      0.162162      0.333333     0.646341     0.454545  ...  0.0  0.0   
69270      0.405405      0.564103     0.536585     0.616162  ...  0.0  0.0   
69271      0.459459      0.717949     0.597561     0.464646  ...  0.0  0.0   
...             ...           ...          ...          ...  ...  ...  ...   
71585      0.054054      0.230769     0.780488     0.363636  ...  0.0  0.0   
71586      0.000000      0.282051     0.756098     0.515152  ...  0.0  0.0   
71587      0.000000      0.179487     0.707317     0.282828  ...  0.0  0.0   
71588      0.108108      0.102564     0.487805     0.434343  ...  0.0  0.0   
71589      0.000000      0.230769     0.792683     0.535354  ...  0.0  0.0   

        91   92   93   94   95   96   97   98  
69267  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  
69268  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  
69269  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
69270  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
69271  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
...    ...  ...  ...  ...  ...  ...  ...  ...  
71585  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  
71586  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  
71587  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  
71588  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
71589  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  

[2323 rows x 116 columns]

In [181]:
def startyfikacja(lista_region_df):
    lista_modeli = []
    lista_x_test = []
    lista_y_test = []
    lista_accu_region = []
    for region in lista_region_df:
        stratified = StratifiedShuffleSplit(n_splits = 4, test_size=200, random_state = 0)
                                            
        df1 = region.copy()
        Rain = df1['RainTomorrow']
        df1.drop(labels=["RainTomorrow",'Location'],axis=1, inplace=True, errors='raise')
        
        for train_index, test_index in stratified.split(df1,Rain):
            x_train = df1.iloc[train_index]
            y_train = Rain.iloc[train_index]
            x_test =  df1.iloc[test_index]
            y_test = Rain.iloc[test_index]

            log = LogisticRegression(random_state=0,solver='lbfgs')
            log.max_iter = 1000
            log.fit(x_train, y_train)
            
            print(log.n_iter_)
            print(log.score(x_test,y_test))
            
            lista_modeli.append(log)
            lista_x_test.append(x_test)
            lista_y_test.append(y_test)
            lista_accu_region.append(log.score(x_test,y_test))
            
    return lista_modeli,lista_x_test,lista_y_test,lista_accu_region

In [182]:
modele,x_test,deszcz_test,lista_accu = startyfikacja(lista_region_df)

[81]
0.85
[82]
0.905
[75]
0.905
[84]
0.9
[85]
0.84
[107]
0.815
[101]
0.78
[86]
0.825
[80]
0.89
[85]
0.895
[74]
0.89
[90]
0.9
[53]
0.955
[71]
0.945
[67]
0.96
[55]
0.93
[82]
0.855
[81]
0.885
[85]
0.86
[89]
0.88
[83]
0.855
[83]
0.87
[71]
0.855
[72]
0.845
[59]
0.87
[71]
0.89
[86]
0.89
[81]
0.87
[98]
0.865
[97]
0.89
[100]
0.88
[100]
0.89
[108]
0.815
[108]
0.82
[91]
0.825
[111]
0.845
[91]
0.91
[98]
0.905
[99]
0.9
[91]
0.875
[61]
0.905
[72]
0.935
[74]
0.935
[63]
0.92
[72]
0.835
[79]
0.84
[62]
0.84
[61]
0.81
[89]
0.835
[87]
0.825
[89]
0.845
[87]
0.835
[83]
0.9
[85]
0.91
[92]
0.91
[77]
0.895
[83]
0.865
[79]
0.895
[72]
0.85
[95]
0.86
[101]
0.86
[96]
0.835
[83]
0.805
[95]
0.83
[47]
0.9
[50]
0.91
[62]
0.915
[50]
0.91
[102]
0.855
[92]
0.86
[92]
0.88
[83]
0.85
[75]
0.845
[77]
0.82
[96]
0.8
[70]
0.815
[75]
0.87
[66]
0.86
[73]
0.865
[72]
0.88
[72]
0.94
[60]
0.945
[72]
0.92
[76]
0.94
[62]
0.92
[73]
0.93
[66]
0.935
[70]
0.93
[75]
0.875
[85]
0.885
[67]
0.82
[77]
0.83
[85]
0.865
[79]
0.87
[80]
0.88
[80]
0

In [183]:
x_test[4]

MinTemp   MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  \
80391  0.587912  0.396088  0.666667       0.517241      0.054054   
79170  0.612637  0.449878  0.000000       0.517241      0.054054   
79154  0.500000  0.374083  0.533333       0.517241      0.351351   
78378  0.541209  0.452323  0.933333       0.517241      0.108108   
80182  0.500000  0.342298  0.000000       0.517241      0.351351   
...         ...       ...       ...            ...           ...   
79229  0.417582  0.303178  0.533333       0.517241      0.351351   
78476  0.370879  0.288509  0.800000       0.517241      0.054054   
80121  0.538462  0.469438  0.000000       0.517241      0.108108   
80287  0.609890  0.535452  0.000000       0.517241      0.702703   
80295  0.500000  0.523227  0.000000       0.517241      0.459459   

       WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  ...  \
80391      0.487179     0.926829     0.515152     0.577039     0.704478  ...   
79170      0.000000     0.560976     0.666667     0.583082     0.620896  ...   
79154      0.487179     0.634146     0.515152     0.858006     0.874627  ...   
78378      0.487179     0.804878     0.616162     0.546828     0.534328  ...   
80182      0.487179     0.634146     0.515152     0.740181     0.677612  ...   
...             ...          ...          ...          ...          ...  ...   
79229      1.000000     0.817073     0.727273     0.528701     0.602985  ...   
78476      0.102564     0.560976     0.606061     0.595166     0.567164  ...   
80121      0.487179     0.743902     0.515152     0.779456     0.755224  ...   
80287      0.487179     0.451220     0.515152     0.371601     0.262687  ...   
80295      0.487179     0.573171     0.515152     0.528701     0.525373  ...   

        89   90   91   92   93   94   95   96   97   98  
80391  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
79170  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
79154  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
78378  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
80182  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
79229  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  
78476  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  
80121  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
80287  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
80295  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  

[200 rows x 114 columns]

In [184]:
len(modele)

196

In [200]:
def sprawdz(modele,x_test,deszcz_test):
    test_size = 200
    XTest = pd.DataFrame()
    YTest = pd.DataFrame()
    lista_wynikow = []
    
    for x in x_test:
        XTest = XTest.append(x.head(test_size))
        
    
    for y in deszcz_test:
        YTest = YTest.append(pd.DataFrame(y).head(test_size))
        
    
    for i in modele:
        lista_wynikow.append(i.score(XTest,YTest))
    return lista_wynikow,XTest,YTest

In [201]:
Wyniki,X_test_glob,Y_test_glob = sprawdz(modele,x_test,deszcz_test)     #Wynikowe df z score danych modeli
Wyniki = pd.DataFrame(Wyniki)
Wyniki.sort_values(0)

0
125  0.799311
127  0.800102
119  0.800918
118  0.801786
126  0.802730
..        ...
56   0.856122
106  0.856352
57   0.856837
59   0.857066
105  0.857092

[196 rows x 1 columns]

In [187]:
Wyniki_region = pd.DataFrame(lista_accu)
Wyniki_region.sort_values(0)

0
109  0.745
6    0.780
110  0.790
135  0.800
74   0.800
..     ...
14   0.960
164  0.960
194  0.965
192  0.970
193  0.970

[196 rows x 1 columns]

In [188]:
index = Wyniki.idxmax() #najwyższe acc na globalnym zbiorze
index_region = Wyniki_region.idxmax() #najwyzsze acc na zbiorze z regionu
display(index)
display(index_region)

0    105
dtype: int64

0    192
dtype: int64

Wyzsza skutecznosc zostala osiagnieta na zbiorze z regionu, dlatego ze na tym był trenowany,
dodatkowo model o najwyzszej skutecznosci w calym kraju, nie jest najlepszy regionalnie, co jest rozsądne, bo byłby zbyt dostosowany do regionu, chyba, że byłby to bardzo stabilny pogodowo region. 

In [203]:
confusion_matrix(Y_test_glob, modele[105].predict(X_test_glob)) #confusion matrix 
#najlepszego globalnie klasyfikatora

array([[ 1153,  4891],
       [  711, 32445]], dtype=int64)

Skutecznosc na wysokosci 85%, jest dość satysfakcjonująca jak na statyczność danych dostarczanych do klasyfikatora, posiadajac taki zbiór danych jesteśmy w stanie dość swobodnie "zgadywać" wystąpienie deszczu w dniu następnym, statystycznie w roku zmokniemy tylko 54 razy :). 